File: PandorasBox/Ahorn/entities/lamp.jl
module PandorasBoxLamp

using ..Ahorn, Maple

@mapdef Entity "pandorasBox/lamp" Lamp(x::Integer, y::Integer, flag::String="", baseColor::String="White", lightColor::String="White")

const colors = sort(collect(keys(Ahorn.XNAColors.colors)))

const placements = Ahorn.PlacementDict(
    "Lamp (Pandora's Box)" => Ahorn.EntityPlacement(
        Lamp,
        "point"
    )
)

Ahorn.editingOptions(entity::Lamp) = Dict{String, Any}(
    "baseColor" => colors,
    "lightColor" => colors
)

function getColor(color)
    if haskey(Ahorn.XNAColors.colors, color)
        return Ahorn.XNAColors.colors[color]

    else
        try
            return ((Ahorn.argb32ToRGBATuple(parse(Int, replace(color, "#" => ""), base=16))[1:3] ./ 255)..., 1.0)

        catch

        end
    end

    return (1.0, 1.0, 1.0, 1.0)
end

gemTexture = "objects/pandorasBox/lamp/idle0"
baseTexture = "objects/pandorasBox/lamp/base"

function Ahorn.selection(entity::Lamp)
    x, y = Ahorn.position(entity)

    return Ahorn.getSpriteRectangle(gemTexture, x, y)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::Lamp, room::Maple.Room)
    colorRaw = get(entity, "baseColor", "White")
    color = getColor(colorRaw)

    Ahorn.drawSprite(ctx, baseTexture, 0, 0, tint=color)
    Ahorn.drawSprite(ctx, gemTexture, 0, 0)
end

end

File: PandorasBox/Loenn/entities/lamp.lua
local drawableSprite = require("structs.drawable_sprite")

local lamp = {}

lamp.name = "pandorasBox/lamp"
lamp.depth = 5
lamp.fieldInformation = {
    baseColor = {
        fieldType = "color",
        allowXNAColors = true,
    },
    lightColor = {
        fieldType = "color",
        allowXNAColors = true,
    },
    lightStartRadius = {
        fieldType = "integer",
    },
    lightEndRadius = {
        fieldType = "integer",
    },
}
lamp.placements = {
    name = "lamp",
    data = {
        inverted = false,
        flag = "",
        baseColor = "White",
        lightColor = "White",
        lightMode = "Smooth",
        lightStartRadius = 48,
        lightEndRadius = 64
    }
}

local gemTexture = "objects/pandorasBox/lamp/idle0"
local baseTexture = "objects/pandorasBox/lamp/base"

function lamp.sprite(room, entity)
    local colorRaw = entity.baseColor or "White"

    local baseSprite = drawableSprite.fromTexture(baseTexture, entity)
    local gemSprite = drawableSprite.fromTexture(gemTexture, entity)

    baseSprite:setColor(colorRaw)

    return {
        baseSprite,
        gemSprite
    }
end

return lamp